src/main/java/org/heigit/ohsome/ohsomeapi/executor/ElementsRequestExecutor.java

```diff
@@ -299,52 +300,59 @@ public static void extractFullHistory(ElementsGeometry elemGeom,
           inputProcessor.getRequestUrlIfGetRequest(servletRequest));
     }
     DataResponse osmData = new DataResponse(new Attribution(URL, TEXT), Application.API_VERSION,
         metadata, "FeatureCollection", Collections.emptyList());
-    try (Stream<Feature> contributionStream = contributionPreResult.stream();
-        Stream<Feature> snapshotStream = snapshotPreResult.stream()) {
-      exeUtils.streamElementsResponse(servletResponse, osmData, true, snapshotStream,
-          contributionStream);
+    MapReducer<Feature> snapshotPreResult = null;
+    if (!onlyModifications) {
+      // handles cases where valid_from = t_start, valid_to = t_end; i.e. non-modified data
+      snapshotPreResult = mapRedSnapshot.groupByEntity().filter(snapshots -> snapshots.size() == 2)
+          .filter(snapshots -> snapshots.get(0).getGeometry() == snapshots.get(1).getGeometry()
+              && snapshots.get(0).getEntity().getVersion() == snapshots.get(1).getEntity()
+                  .getVersion())
+          .map(snapshots -> snapshots.get(0)).flatMap(snapshot -> {
+            Map<String, Object> properties = new TreeMap<>();
+            OSMEntity entity = snapshot.getEntity();
+            if (includeOSMMetadata) {
+              properties.put("@lastEdit", entity.getTimestamp().toString());
+            }
+            Geometry geom = snapshot.getGeometry();
+            if (unclippedGeometries) {
+              geom = snapshot.getGeometryUnclipped();
+            }
+            properties.put("@snapshotTimestamp",
+                TimestampFormatter.getInstance().isoDateTime(snapshot.getTimestamp()));
+            properties.put("@validFrom", startTimestamp);
+            properties.put("@validTo", endTimestamp);
+            boolean addToOutput;
+            if (processingData.containsSimpleFeatureTypes()) {
+              addToOutput = utils.checkGeometryOnSimpleFeatures(geom, simpleFeatureTypes);
+            } else if (requiresGeometryTypeCheck) {
+              addToOutput = filterExpression.applyOSMGeometry(entity, geom);
+            } else {
+              addToOutput = true;
+            }
+            if (addToOutput) {
+              return Collections.singletonList(exeUtils.createOSMFeature(entity, geom, properties,
+                  keysInt, includeTags, includeOSMMetadata, elemGeom));
+            } else {
+              return Collections.emptyList();
+            }
+          }).filter(Objects::nonNull);
+      try (Stream<Feature> contributionStream = contributionPreResult.stream();
+          Stream<Feature> snapshotStream = snapshotPreResult.stream()) {
+        exeUtils.streamElementsResponse(servletResponse, osmData, true, snapshotStream,
+            contributionStream);
+      }
+    } else {
+      try (Stream<Feature> contributionStream = contributionPreResult.stream()) {
+        exeUtils.streamElementsResponse(servletResponse, osmData, true, null,
+            contributionStream);
+      }
     }
```

Comment:
code duplication
Stream implements Autoclosable but it is generally not necessary to close streams at all. You only need to close streams that use IO resources, which is this case they do not.

You can avoid this duplication here by moving the declaration of contributionStream and snapshotStream before the if block

```
Stream<Feature> contributionStream = contributionPreResult.stream();
Stream<Feature> snapshotStream = null;
if (!onlyModifications) {
...
  snapshotStream = snapshotPreResult.stream();
}
 exeUtils.streamElementsResponse(servletResponse, osmData, true, 
   snapshotStream,
   contributionStream);
```

Reply:
Why do you think this is not necessary to (auto)close the stream here? At least in some backends, the closing of the result stream would actually close some otherwise left open IO resources, e.g. here: https://github.com/GIScience/oshdb/blob/29c29056615569a45195deac60355b238ad3df7e/oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/backend/MapReducerIgniteScanQuery.java#L545-L547

Reply:
That's a good point, and I would guess, that the majority of users would never close the stream like me :-).
So we should write a warning in the doc that this is a resource stream and should be close.

For this request I suggest,as it was to pull the `snapshotPerResult` for the if block again and check it in the try-with-resources later
```
try (Stream<Feature> snapshotStream = (snapshotPreResult != null)?snapshotPreResult.stream():null;
     Stream<Feature> contributionStream = contributionPreResult.stream()) {
  exeUtils.streamElementsResponse(servletResponse, osmData, true, snapshotStream, contributionStream);
}
```

Reply:
Related to the MapReducerIgniteScanQuery I found a nice article, that the flatMap operation will close the inner stream in any case.
https://rmannibucau.metawerx.net/post/did-you-think-about-closing-your-java-stream section "java.util.stream, close and flatMap"
And luckily this is exactly how we handle the returning Stream of `mapStreamOnIgniteCache`

https://github.com/GIScience/oshdb/blob/29c29056615569a45195deac60355b238ad3df7e/oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/backend/MapReducerIgniteScanQuery.java#L198

Reply:
thanks, I didn't know about this! good to know. I have to say though that I find it quite strange that they did implement this "automatic closing of a stream when it is consumed" for `flatMap`'ed streams, but not regular streams. :shrug: